storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java

```diff
@@ -445,7 +445,7 @@ public FetchDataInfo read(long startOffset, int maxSize, long maxPosition, boole
             adjustedMaxSize = Math.max(maxSize, startOffsetAndSize.size);
 
         // return a log segment but with zero size in the case below
-        if (adjustedMaxSize == 0)
+        if (adjustedMaxSize == 0 || maxPosition == -1)
```

Comment:
1. It's better to use Option[Long] for maxPosition. This also makes it consistent with maxSize for which negative values are invalid.
2. Could we adjust the comment above?

Reply:
LogSegment is in Java and using Optional as a method parameter in LogSegment#read shows warnings in the intelliJ and require some refactoring, so went with the negative value approach:

```
'Optional<Long>' used as type for parameter 'maxPosition' 
```